tests/self/macro.bt

```diff
@@ -0,0 +1,391 @@
+macro inc($x)
+{
+  $x += 1;
+}
+
+test:macro_mutates_variable
+{
+  $a = 1;
+  inc($a);
+  if ($a != 2) {
+    return 1;
+  }
+}
+
+macro inc2($x)
+{
+  $x += 1;
+  $x
+}
+
+test:macro_assigns_to_last_expression
+{
+  $a = 1;
+  $b = inc2($a);
+  if ($a != 2 || $b != 2) {
+    return 1;
+  }
+}
+
+macro set_nonscalar(@x)
+{
+  @x[1] = 2;
+}
+
+test:macro_mutates_nonscalar_map
+{
+  @a[1] = 1;
+  set_nonscalar(@a);
+  if (@a[1] != 2) {
+    return 1;
+  }
+  delete(@a[1]);
+}
+
+macro set_scalar(@x)
+{
+  @x = 2;
+}
+
+test:macro_mutates_scalar_map
+{
+  @b = 1;
+  set_scalar(@b);
+  if (@b != 2) {
+    return 1;
+  }
+  delete(@b);
+}
+
+macro add(x)
+{
+  let $y : typeof(x);
+  if true {
+    $y = 1;
+  }
+  x + $y
+}
+
+test:macro_renames_internal_variables
+{
+  if (add(1) != 2) {
+    return 1;
+  }
+}
+
+macro get(@x)
+{
+  1 + @x[1]
+}
+
+test:macro_reads_passed_map
+{
+  @c[1] = 1;
+  if (get(@c) != 2) {
+    return 1;
+  }
+  delete(@c[1]);
+}
+
+macro add1(x)
+{
+  x + 1
+}
+macro add2(y)
+{
+  y + add1(y)
+}
+macro add3(z)
+{
+  z + add2(z)
+}
+
+test:macro_calls_other_macros_expressions
+{
+  if (add3(1) != 4) {
+    return 1;
+  }
+}
+
+macro add1s($x)
+{
+  $x += 1;
+}
+macro add2s($y)
+{
+  add1s($y);
+  $y += 1;
+}
+macro add3s($z)
+{
+  add2s($z);
+  $z += 1;
+}
+
+test:macro_calls_other_macros_statements
+{
+  $a = 1;
+  add3s($a);
+  if ($a != 4) {
+    return 1;
+  }
+}
+
+macro add2_ord(x)
+{
+  x + add1_ord(x)
+}
+macro add3_ord(x)
+{
+  x + add2_ord(x)
+}
+macro add1_ord(x)
+{
+  x + 1
+}
+
+test:macro_definition_order_does_not_matter
+{
+  if (add3_ord(1) != 4) {
+    return 1;
+  }
+}
+
+macro add_one(x)
+{
+  x + 1
+}
+
+test:macro_accepts_arbitrary_expressions_without_vars
+{
+  $a = 1;
+  if (add_one(1 + 1) != 3) {
+    return 1;
+  }
+}
+
+test:macro_accepts_arbitrary_expressions_with_vars
+{
+  $a = 1;
+  if (add_one($a + 1) != 3) {
+    return 1;
+  }
+}
+
+macro add_two(x)
+{
+  x + 2
+}
+macro add_one_nested(x)
+{
+  add_two(x + 1) + 1
+}
+
+test:macro_accepts_arbitrary_expressions_nested
+{
+  $a = 1;
+  if (add_one_nested($a + 1) != 6) {
+    return 1;
+  }
+}
+
+test:macro_accepts_arbitrary_expressions_with_maps
+{
+  @d = 1;
+  if (add_one(@d + 1) != 3) {
+    return 1;
+  }
+  delete(@d);
+}
+
+macro inc_multi($x)
+{
+  $x += 1;
```

Comment:
Just pointing out - we have 3 macros with exactly the same body. I get that the reason is so that it's easier to read the tests but still feels superfluous 😅